### PR TITLE
[MM-48538] `/apps debug logs` to stream plugin logs to a Mattermost channel

### PR DESCRIPTION
**Pull request by levb**
#### Summary
Added the ability to stream all `mattermost-plugin-apps` logs to a channel. Use `/apps debug logs [--create-channel true | --channel <channel>] [--level] [--json]` to control.

The "core" change: https://github.com/mattermost/mattermost-plugin-apps/pull/394/files#diff-9c4561950e563d1167e9e955b5ab4c6ce45184275371b672a33bc16b1ccdcfb9R96

Looks like this, 
image

Also, some cleanup, including unused mocks.

#### Ticket Link
[MM-48538](https://mattermost.atlassian.net/browse/MM-48538)

### CODE REVIEW

File: server/proxy/service.go
Comment (javaguirre) [inline server/proxy/service.go:47]: should we remove this?

File: server/proxy/service.go
Comment (javaguirre) [inline server/proxy/service.go:124]: Idem

File: server/proxy/service.go
Comment (javaguirre) [reply server/proxy/service.go:47]: Or do you plan to add logs here somewhere?

**Review by javaguirre — APPROVED** [review]
LGTM, just commented out details

**Comment by asatkinson** [comment]
This is awesome.  Here is a JIRA ticket in the developer experience EPIC
https://mattermost.atlassian.net/browse/MM-48538?atlOrigin=eyJpIjoiOTBlNDFiYjJiOTk5NDI1NjgzZGI4NmE1NjY3OWQxNGQiLCJwIjoiaiJ9

**Review by asatkinson — APPROVED** [review]
This is awesome. I created a JIRA ticket in the developer experience EPIC.  We can move it if desired.
https://mattermost.atlassian.net/browse/MM-48538?atlOrigin=eyJpIjoiOTBlNDFiYjJiOTk5NDI1NjgzZGI4NmE1NjY3OWQxNGQiLCJwIjoiaiJ9

- This seems like a lot of changes.  Please make sure you have good review on this.  If this PR slips due to other priorities that is OK, but it would be nice to have.
- Let me know if you need me to queue up documentation for this.  We can ship this undocumented for a while if need be.

**Review by mickmister — APPROVED** [review]

File: server/config/service.go
Comment (hanzei) [inline server/config/service.go:226]: When this line gets called, `s.mattermostConfig` has not been refreshed. This means a change in the dev mode does not get picked up. Also, the logger used in `newInitializedConfig` will not use the updated `LogConfig`.

File: server/builtin/debug_logs.go
Comment (hanzei) [inline server/builtin/debug_logs.go:64]: This seems off.

File: server/builtin/debug_logs.go
Comment (hanzei) [inline server/builtin/debug_logs.go:164]: If we use `ch.Name` here, we get an autolink.

File: server/httpin/service.go
Comment (hanzei) [inline server/httpin/service.go:44]: Good cleanup

File: server/plugin.go
Comment (hanzei) [inline server/plugin.go:58]: In order to reduce the number of different loggers used, should a `PluginLogger` be used here?
```golang
log := utils.NewPluginLogger(mm, nil)
```

Once `p.conf` is setup, `log` gets reassigned `log = p.conf.NewBaseLogger()`

File: server/proxy/synchronize.go
Comment (hanzei) [inline server/proxy/synchronize.go:26]:
```suggestion
	r := p.NewIncomingRequest().WithCtx(ctx)
```

File: server/proxy/synchronize.go
Comment (hanzei) [inline server/proxy/synchronize.go:28]: Should we pass a `Logger` to `SynchronizeInstalledApps`?

File: server/config/service.go
Comment (levb) [reply server/config/service.go:226]: `configLogger` is the logger to use while handling the configuration change itself, so it is appropriate that it uses the "old" configuration for itself. Also, FWIW, it is set to a "real" value only when called the first time in `OnActivate`, afterwards it's always `nil`.

All other config changes should be picked up by `newInitializedConfig`, or am I missing something else?

File: server/proxy/synchronize.go
Comment (levb) [reply server/proxy/synchronize.go:28]: 1/5 it's fine as is, since it creates a new request with its own logger and uses it. The old explicit log initialization is already gone.

File: utils/plugin_logger.go
Comment (levb) [reply utils/plugin_logger.go:43]: 1/5 There are lots of improvements as to _what_ we actually log, would prefer a separate PR.

File: utils/plugin_logger.go
Comment (levb) [reply utils/plugin_logger.go:116]: 1/5 we should have another PR that ensures "minimum viable text" in all log messages, and meaningful JSON improvements; not here.

File: utils/plugin_logger.go
Comment (levb) [reply utils/plugin_logger.go]: Updated; is this what you had in mind?

**Comment by levb** [comment]
Added a draft docs PR for @asatkinson to polish.

**Comment by javaguirre** [comment]
Screenshot 2022-12-05 at 12 40 16
Screenshot 2022-12-05 at 12 40 03

Not sure if I'm doing something wrong, but the command doesn't appear, I'm gonna try to use this along with the `apps settings` PR in case the developer mode is not set up correctly.

**Comment by javaguirre** [comment]
It seems the QA is done for this one though, moving to another one.

File: server/config/service.go
Comment (hanzei) [reply server/config/service.go]: By changing the code to pass a `Logger` to `Reconfigure` the bug has been fixed. :+1:

**Comment by javaguirre** [comment]
Need to test with other cases, but overall it's working well. I can use it, the channel configuration went well and I receive logs.

Screenshot 2022-12-05 at 13 15 02

**Comment by neflyte** [comment]
What happens if I turn on logging with `--create-channel` twice in a row? Will I get two channels that have the same name? If I have two Apps and I turn on this debugging for both, do I also get two channels with the same name? If so, which channel maps to which App?

**Comment by neflyte** [comment]
Also, how do I disable the logging once it's enabled? In a production scenario, I'd want to be able to turn it on, get the logs i need, and turn it back off.

**Comment by hanzei** [comment]
@levb Heads up that there are merge conflicts to resolve

**Comment by javaguirre** [comment]
> What happens if I turn on logging with `--create-channel` twice in a row? Will I get two channels that have the same name? If I have two Apps and I turn on this debugging for both, do I also get two channels with the same name? If so, which channel maps to which App?
> 

Now It only creates one channel and all apps write to that channel if logs are enabled. If you post `--create-channel` twice, it's the same channel and it's not created twice. 

> Perhaps change `--create-channel` to accept a channel name? A channel would be created if the parameter is non-empty.

You could create a channel manually and then use `--channel` to select it.

0/5 I don't think selecting a channel for creation is really needed, you can do it in two steps.

**Comment by javaguirre** [comment]
> Also, how do I disable the logging once it's enabled? In a production scenario, I'd want to be able to turn it on, get the logs i need, and turn it back off.

Screenshot 2022-12-07 at 09 50 13

In order to use the Apps logging, you need to set `Developer mode` and `Testing Commands` to true. When you go back to Production mode we set these to false, and the logging will stop.

File: utils/plugin_logger.go
Comment (hanzei) [reply utils/plugin_logger.go:45]: Sure, submitted https://github.com/mattermost/mattermost-plugin-apps/pull/410

File: utils/plugin_logger.go
Comment (hanzei) [reply utils/plugin_logger.go:118]: Could you elaborate a bit on your idea? Most of our log messages are quite generic. The value comes from the context i.e. the fields.

**Comment by hanzei** [comment]
@levb Heads up that I merged `master` into your branch. Please double check [`83d1308` (#394)](https://github.com/mattermost/mattermost-plugin-apps/pull/394/commits/83d13085a28c458ca07a3d91bd8ad78c04aff5e5).

**Review by hanzei — APPROVED** [review]

**Comment by hanzei** [comment]
@javaguirre @DHaussermann The PR is ready for testing.

**Comment by javaguirre** [comment]
I will test the following cases:

- [ ] Activate logs creating a new channel
- [ ] Activate logs selecting an existing channel
- [ ] Receive logs on a channel without the `--json` flag.
- [ ] Receive logs on a channel with the `--json` flag.
- [ ] Deactivating developer mode and/or testing commands with stop logging
- [ ] If we create a channel and it was created, it's not created twice

**Comment by javaguirre** [comment]
If I select a team and I activate logs for an existing channel within that team, the logs are shown in a different team (the last one created probably if the channel has the same name) @levb @hanzei 

If I use `--create-channel` on a team channel, the logs are shown in the channel `DEBUG: Apps` of that same team as intended.

https://user-images.githubusercontent.com/488556/207268974-160f19af-8ed7-482f-a053-1c7266ca7261.mov

**Review by javaguirre — CHANGES_REQUESTED** [review]
We need to fix this use case

https://github.com/mattermost/mattermost-plugin-apps/pull/394#issuecomment-1347948166

**Comment by hanzei** [comment]
Taking a look, @javaguirre

**Comment by levb** [comment]
@hanzei @javaguirre heads up: I am changing this a bit to combine with the Settings command.

**Comment by hanzei** [comment]
Good catch, @javaguirre. :+1: That is indeed a bug in the framework: https://mattermost.atlassian.net/browse/MM-49043.

Given that  @levb is migrating the command into the settings form, it won't block merging this PR.

**Comment by levb** [comment]
@javaguirre since your issue has been filed as a separate ticket, can you approve and unblock this PR?

**Review by javaguirre — APPROVED** [review]

**Pull request merged**
